src/lib/pickles/pickles.ml

```diff
@@ -729,6 +729,139 @@ module Make_str (_ : Wire_types.Concrete) = struct
+                    ; feature_flags = Plonk_types.Features.none_bool
+                    ; prevs = [ self; self ]
+                    ; main =
+                        (fun { public_input = () } ->
+                          let prev direction =
+                            exists Field.typ ~request:(fun () ->
+                                Recursive_input direction )
+                          in
+                          let prev_proof direction =
+                            exists (Typ.prover_value ()) ~request:(fun () ->
+                                Recursive_proof direction )
+                          in
+                          let is_base_case =
+                            exists Boolean.typ ~request:(fun () -> Is_base_case)
+                          in
+                          let proof_must_verify = Boolean.not is_base_case in
+                          let self =
+                            Field.(
+                              if_ is_base_case ~then_:one
+                                ~else_:(prev Left + prev Right))
+                          in
+                          Promise.return
+                            { Inductive_rule.previous_proof_statements =
+                                [ { public_input = prev Left
+                                  ; proof = prev_proof Left
+                                  ; proof_must_verify
+                                  }
+                                ; { public_input = prev Right
+                                  ; proof = prev_proof Right
+                                  ; proof_must_verify
+                                  }
+                                ]
+                            ; public_output = self
+                            ; auxiliary_output = ()
+                            } )
+                    }
+                  ] ) )
+
+        module Proof = (val p)
+
+        let fib5 =
+          let s_neg_one = Field.Constant.(negate one) in
+          let b_neg_one : (Nat.N2.n, Nat.N2.n) Proof0.t =
+            Proof0.dummy Nat.N2.n Nat.N2.n Nat.N2.n ~domain_log2:15
+          in
+          let s0, (), b0 =
+            Common.time "tree b0" (fun () ->
+                Promise.block_on_async_exn (fun () ->
+                    step
+                      ~handler:
+                        (handler true (s_neg_one, b_neg_one)
+                           (s_neg_one, b_neg_one) )
+                      () ) )
+          in
+          assert (Field.Constant.(equal one) s0) ;
+          Or_error.ok_exn
+            (Promise.block_on_async_exn (fun () ->
+                 Proof.verify_promise [ (s0, b0) ] ) ) ;
+
+          let rec fib_proof_actual n =
+            if n <= 1 then (s0, b0)
+            else
+              let sn, (), bn =
+                Promise.block_on_async_exn (fun () ->
+                    step
+                      ~handler:
+                        (handler false (fib_proof (n - 1)) (fib_proof (n - 2)))
```

Comment:
Nit: I assume this is the intention, but because the Fibonnaci logic only lives here -- outside of the circuit -- then we can prove any number.
If that is the intention, could you add a comment explaining this?

Reply:
Concretely, for any field element `f`, we can decompose it into bits, create proofs of the powers of 2 by running `step` repeatedly on the previous proofs, and then build that field element by combining the proofs corresponding to the bits that we want.

Reply:
Is there a way I could constraint this circuit so it only accept fibs?

Reply:
Do I need to modify the Request like this?
![image](https://github.com/user-attachments/assets/fd84d30d-baa3-489d-b628-f41934f87896)